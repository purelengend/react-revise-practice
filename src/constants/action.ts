export const URL_ACTION = {
  SORT: "SORT",
  ORDER: "ORDER",
  PAGE: "PAGE",
  FILTER: "FILTER",
} as const;
