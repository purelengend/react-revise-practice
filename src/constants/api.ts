export const API = {
  STUDENT: "VITE_STUDENT_API",
  IMAGE: "VITE_UPLOAD_IMAGE_API",
  IMGBB_KEY: "VITE_IMGBB_API_KEY",
} as const;
