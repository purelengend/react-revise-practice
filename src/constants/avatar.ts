export const DEFAULT_AVATAR_URL = "/assets/images/user-photo.svg";
