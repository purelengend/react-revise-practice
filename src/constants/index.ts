export * from "./axios";
export * from "./query";
export * from "./role";
export * from "./avatar";
