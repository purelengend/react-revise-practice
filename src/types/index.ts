export * from "./student";
export * from "./role";
export * from "./sidebar";
export * from "./sort";
