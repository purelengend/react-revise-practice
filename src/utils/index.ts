export * from "./test-utils";
export * from "./config";
export * from "./customToast";
