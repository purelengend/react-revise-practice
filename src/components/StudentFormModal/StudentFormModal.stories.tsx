import type { Meta, StoryObj } from "@storybook/react";
import { Center } from "@chakra-ui/react";
import { useArgs } from "@storybook/preview-api";

// Components
import StudentFormModal from ".";

// Constants
import { DEFAULT_STUDENT_AVATAR_URL } from "@/constants";

const meta: Meta<typeof StudentFormModal> = {
  title: "Components/StudentFormModal",
  component: StudentFormModal,
  decorators: [
    (Story) => (
      <Center style={{ width: "100vh" }}>
        <Story />
      </Center>
    ),
  ],
  argTypes: {
    isOpen: {
      description: "Determine whether the modal is open or not",
    },
    onClose: {
      description: "Function to close the modal",
    },
    onSubmit: {
      description: "Function to submit the form",
    },
    isMutating: {
      description: "Determine whether the form is mutating or not",
    },
    student: {
      description: "Student data of the form",
    },
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof StudentFormModal>;

export const Primary: Story = {
  args: {
    isOpen: true,
    onSubmit: () => {},
    isMutating: false,
    student: {
      id: "1",
      name: "John Doe",
      email: "john.doe@gmail.com",
      avatarUrl: DEFAULT_STUDENT_AVATAR_URL,
      dateOfAdmission: Date.now(),
      phone: "0905150899",
    },
  },
  render: function Render(args) {
    const [, updateArgs] = useArgs<typeof args>();

    const onClose = () => {
      updateArgs({ isOpen: false });
    };

    return <StudentFormModal {...args} onClose={onClose} />;
  },
};
