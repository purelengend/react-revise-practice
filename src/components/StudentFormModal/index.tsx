import {
  Box,
  Button,
  Center,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Image,
  Input,
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
} from "@chakra-ui/react";
import { memo, useCallback, useState } from "react";
import { Controller, useForm } from "react-hook-form";
import { valibotResolver } from "@hookform/resolvers/valibot";

// Components
import { AttachmentIcon } from "../common/Icons";

// Types
import { Student } from "@/types";

// Schema
import { StudentSchema } from "@/schema";

// Hooks
import { useImage } from "@/hooks";
import { DEFAULT_STUDENT_AVATAR_URL } from "@/constants";
import { formatPhoneNumber, onlyNumberKeyDown } from "@/utils";

export type StudentFormModalProps = {
  isOpen: boolean;
  isMutating: boolean;
  student: Student;
  onClose: () => void;
  onSubmit: (data: Student) => void;
};

const StudentFormModal = memo(
  ({
    isOpen,
    isMutating,
    onClose,
    student,
    onSubmit,
  }: StudentFormModalProps) => {
    const [selectedImage, setSelectedImage] = useState<File>();

    const { mutate: uploadImage, isPending: isUploadingImage } = useImage();

    const {
      control,
      handleSubmit,
      setValue,
      formState: { errors, isDirty },
    } = useForm<Student>({
      defaultValues: student,
      resolver: valibotResolver(StudentSchema),
      mode: "onBlur",
    });

    const handleImageUpload = useCallback(
      async (e: React.ChangeEvent<HTMLInputElement>) => {
        if (e.target.files) {
          const imageFile = e.target.files[0];

          const imageFormData = new FormData();

          imageFormData.append("image", imageFile);

          uploadImage(imageFormData, {
            onSuccess: (response) => {
              const imageUrl = response.data.data.url;

              setValue("avatarUrl", imageUrl, { shouldDirty: true });
            },
            onError: () => {
              setSelectedImage(undefined);
            },
          });

          setSelectedImage(imageFile);
        }
      },
      [setValue, uploadImage],
    );

    return (
      <Modal
        autoFocus={false}
        returnFocusOnClose={false}
        isOpen={isOpen}
        onClose={onClose}
      >
        <ModalOverlay />
        <ModalContent
          bg="white"
          px={{
            base: 0,
            md: 15.75,
          }}
          minW={{
            base: "auto",
            md: 169,
          }}
        >
          <ModalHeader
            textAlign="center"
            py={9}
            borderBottom="1px solid"
            borderBottomColor="white.300"
            fontSize="2xl"
          >
            {student.id ? "Edit student" : "Add student"}
          </ModalHeader>
          <form id="#student-form" noValidate onSubmit={handleSubmit(onSubmit)}>
            <ModalBody py={9}>
              <FormControl
                isInvalid={!!errors}
                display="flex"
                flexDir="column"
                gap={5}
                opacity={isMutating ? 0.5 : 1}
              >
                <Center>
                  <Box bg="gray.400" borderRadius="50%" pos="relative">
                    {selectedImage ? (
                      <Image
                        boxSize={32}
                        objectFit="cover"
                        borderRadius="50%"
                        opacity={isUploadingImage ? 0.5 : 1}
                        src={URL.createObjectURL(selectedImage)}
                        fallbackSrc={DEFAULT_STUDENT_AVATAR_URL}
                      />
                    ) : (
                      <Image
                        boxSize={32}
                        objectFit="cover"
                        borderRadius="50%"
                        src={student.avatarUrl}
                        fallbackSrc={DEFAULT_STUDENT_AVATAR_URL}
                      />
                    )}
                    <FormLabel
                      htmlFor="#avatar"
                      pos="absolute"
                      right={0}
                      bottom={0}
                      aria-label="upload-image"
                    >
                      <AttachmentIcon />
                    </FormLabel>
                    <Input
                      id="#avatar"
                      name="avatar"
                      type="file"
                      aria-label="avatar"
                      isDisabled={isMutating || isUploadingImage}
                      hidden
                      onChange={handleImageUpload}
                    />
                    <Controller
                      control={control}
                      name="avatarUrl"
                      render={({ field }) => (
                        <Input id="#avatarUrl" {...field} hidden />
                      )}
                    />
                  </Box>
                </Center>

                <Controller
                  control={control}
                  name="name"
                  render={({ field, formState: { errors } }) => (
                    <Box>
                      <FormLabel htmlFor="#name">Name</FormLabel>
                      <Input
                        {...field}
                        id="#name"
                        variant="filled"
                        isInvalid={!!errors.name}
                        isDisabled={isMutating}
                      />
                      {errors.name && (
                        <FormErrorMessage>
                          {errors.name.message}
                        </FormErrorMessage>
                      )}
                    </Box>
                  )}
                />

                <Controller
                  control={control}
                  name="email"
                  render={({ field, formState: { errors } }) => (
                    <Box>
                      <FormLabel htmlFor="#email">Email</FormLabel>
                      <Input
                        {...field}
                        id="#email"
                        type="email"
                        variant="filled"
                        isInvalid={!!errors.email}
                        isDisabled={isMutating}
                      />
                      {errors.email && (
                        <FormErrorMessage>
                          {errors.email.message}
                        </FormErrorMessage>
                      )}
                    </Box>
                  )}
                />

                <Controller
                  control={control}
                  name="phone"
                  render={({
                    field: { onChange, ...rest },
                    formState: { errors },
                  }) => (
                    <Box>
                      <FormLabel htmlFor="#phone">Phone</FormLabel>
                      <Input
                        {...rest}
                        id="#phone"
                        type="text"
                        variant="filled"
                        maxLength={12}
                        isInvalid={!!errors.phone}
                        isDisabled={isMutating}
                        onChange={formatPhoneNumber(onChange)}
                        onKeyDown={onlyNumberKeyDown}
                      />
                      {errors.phone && (
                        <FormErrorMessage>
                          {errors.phone.message}
                        </FormErrorMessage>
                      )}
                    </Box>
                  )}
                />

                <Controller
                  control={control}
                  name="id"
                  render={({ field }) => (
                    <Input {...field} id="#id" type="text" hidden />
                  )}
                />

                <Controller
                  control={control}
                  name="dateOfAdmission"
                  render={({ field }) => (
                    <Input
                      {...field}
                      id="#admission-date"
                      type="number"
                      hidden
                    />
                  )}
                />
              </FormControl>
            </ModalBody>

            <ModalFooter display="flex" gap={5}>
              <Button
                type="button"
                w={30}
                px={7.5}
                py={4}
                onClick={onClose}
                bg="blackAlpha.100"
                _hover={{
                  bg: "blackAlpha.300",
                }}
                isDisabled={isMutating || isUploadingImage}
              >
                Cancel
              </Button>
              <Button
                type="submit"
                w={30}
                px={7.5}
                py={4}
                bg="yellow.200"
                _hover={{
                  bg: "orange.400",
                }}
                isDisabled={isMutating || isUploadingImage || !isDirty}
              >
                Submit
              </Button>
            </ModalFooter>
          </form>
        </ModalContent>
      </Modal>
    );
  },
);

export default StudentFormModal;
