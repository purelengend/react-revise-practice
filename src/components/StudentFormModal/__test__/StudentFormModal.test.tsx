import { screen, renderHook } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useDisclosure } from "@chakra-ui/react";

// Component
import StudentFormModal from "..";

// Types
import { Student } from "@/types";

// Constants
import { FORM_ERROR } from "@/constants";

// Utils
import { customRender } from "@/utils";

describe("StudentFormModal test cases", () => {
  const mockOnSubmit = jest.fn();

  window.URL.createObjectURL = jest.fn();

  const {
    result: {
      current: { isOpen, onClose },
    },
  } = renderHook(() =>
    useDisclosure({
      defaultIsOpen: true,
    }),
  );

  const mockData: Student = {
    id: "",
    name: "",
    email: "",
    phone: "",
    dateOfAdmission: 0,
    avatarUrl: "",
  };

  const mockDataWithId: Student = {
    id: "0",
    name: "valid",
    email: "valid@gmail.com",
    phone: "1234567890",
    dateOfAdmission: 0,
    avatarUrl: "mock",
  };

  const setup = (data: Student) => {
    userEvent.setup();
    return customRender(
      <StudentFormModal
        isOpen={isOpen}
        onClose={onClose}
        onSubmit={mockOnSubmit}
        isMutating={false}
        student={data}
      />,
    );
  };
  it("should render correctly", () => {
    const { container } = setup(mockData);

    expect(container).toMatchSnapshot();
  });

  it('should display the banner text "Edit student" when the input data has a valid ID', async () => {
    setup(mockDataWithId);

    const banner = screen.getByRole("banner");

    expect(banner.textContent).toEqual("Edit student");
  });

  it("should prevent submit when data is invalid", async () => {
    setup(mockData);

    const submitBtn = screen.getByRole("button", {
      name: /submit/i,
    });

    expect(submitBtn).toHaveAttribute("disabled");
  });

  it("should handle upload image", async () => {
    setup(mockData);

    const uploadInput = screen.getByLabelText("avatar");

    const file = new File(["blob"], "chucknorris.png", { type: "image/png" });

    await userEvent.upload(uploadInput, file);
  });

  it("should display error message when typing invalid data", async () => {
    setup(mockData);

    const nameInput = screen.getByRole("textbox", {
      name: /name/i,
    });

    const emailInput = screen.getByRole("textbox", {
      name: /email/i,
    });

    const phoneInput = screen.getByRole("textbox", {
      name: /phone/i,
    });

    await userEvent.type(nameInput, "a");

    await userEvent.tab();

    await userEvent.type(emailInput, "a");

    await userEvent.tab();

    await userEvent.type(phoneInput, "1");

    await userEvent.tab();

    const nameError = screen.getByText(new RegExp(FORM_ERROR.NAME, "i"));

    const emailError = screen.getByText(new RegExp(FORM_ERROR.EMAIL, "i"));

    const phoneError = screen.getByText(new RegExp(FORM_ERROR.PHONE, "i"));

    expect(nameError.textContent).toEqual(FORM_ERROR.NAME);

    expect(emailError.textContent).toEqual(FORM_ERROR.EMAIL);

    expect(phoneError.textContent).toEqual(FORM_ERROR.PHONE);
  });

  it("should invoke submit function when clicking submit button", async () => {
    setup(mockData);

    const nameInput = screen.getByRole("textbox", {
      name: /name/i,
    });

    const emailInput = screen.getByRole("textbox", {
      name: /email/i,
    });

    const phoneInput = screen.getByRole("textbox", {
      name: /phone/i,
    });

    await userEvent.type(nameInput, "valid name");

    await userEvent.tab();

    await userEvent.type(emailInput, "valid@gmail.com");

    await userEvent.tab();

    await userEvent.type(phoneInput, "1234567890");

    await userEvent.tab();

    const submitBtn = screen.getByRole("button", {
      name: /submit/i,
    });

    await userEvent.click(submitBtn);

    expect(mockOnSubmit).toHaveBeenCalled();
  });
});
