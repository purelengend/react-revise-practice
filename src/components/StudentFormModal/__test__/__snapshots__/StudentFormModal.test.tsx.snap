// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentFormModal test cases should render correctly 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
